Drv/Ip/UdpSocket.cpp

```diff
@@ -0,0 +1,165 @@
+
+UdpSocket::~UdpSocket() {
+    FW_ASSERT(m_state);
+    delete m_state;
+}
+
+SocketIpStatus UdpSocket::configureSend(const char* const hostname, const U16 port, const U32 timeout_seconds, const U32 timeout_microseconds) {
+    //Timeout is for the send, so configure send will work with the base class
+    return this->IpSocket::configure(hostname, port, timeout_seconds, timeout_microseconds);
+}
+
+SocketIpStatus UdpSocket::configureRecv(const char* hostname, const U16 port) {
+    this->m_recv_port = port;
+    (void) Fw::StringUtils::string_copy(this->m_recv_hostname, hostname, SOCKET_MAX_HOSTNAME_SIZE);
+    return SOCK_SUCCESS;
+}
+
+
+SocketIpStatus UdpSocket::bind(NATIVE_INT_TYPE fd) {
+    struct sockaddr_in address;
+    FW_ASSERT(-1 != fd);
+    FW_ASSERT(0 != m_recv_port);
+
+    // Set up the address port and name
+    address.sin_family = AF_INET;
+    address.sin_port = htons(m_recv_port);
+    // OS specific settings
+#if defined TGT_OS_TYPE_VXWORKS || TGT_OS_TYPE_DARWIN
+    address.sin_len = static_cast<U8>(sizeof(struct sockaddr_in));
+#endif
+
+    // First IP address to socket sin_addr
+    if (IpSocket::addressToIp4(m_recv_hostname, &address.sin_addr) != SOCK_SUCCESS) {
+        return SOCK_INVALID_IP_ADDRESS;
+    };
+    // UDP (for receiving) requires bind to an address to the socket
+    if (::bind(fd, reinterpret_cast<struct sockaddr*>(&address), sizeof(address)) < 0) {
+        return SOCK_FAILED_TO_BIND;
+    }
+    FW_ASSERT(sizeof(this->m_state->m_addr_recv) == sizeof(address), sizeof(this->m_state->m_addr_recv), sizeof(address));
+    memcpy(&this->m_state->m_addr_recv, &address, sizeof(this->m_state->m_addr_recv));
+    return SOCK_SUCCESS;
+}
+
+SocketIpStatus UdpSocket::openProtocol(NATIVE_INT_TYPE& fd) {
+    SocketIpStatus status = SOCK_SUCCESS;
+    NATIVE_INT_TYPE socketFd = -1;
+    struct sockaddr_in address;
+
+    // Ensure configured for at least send or receive
+    if (m_port == 0 && m_recv_port == 0) {
+        return SOCK_INVALID_IP_ADDRESS; // Consistent with port = 0 behavior in TCP
+    }
+
+    // Acquire a socket, or return error
+    if ((socketFd = ::socket(AF_INET, SOCK_DGRAM, 0)) == -1) {
+        return SOCK_FAILED_TO_GET_SOCKET;
+    }
+
+    // May not be sending in all cases
+    if (this->m_port != 0) {
+        // Set up the address port and name
+        address.sin_family = AF_INET;
+        address.sin_port = htons(this->m_port);
+
+        // OS specific settings
+#if defined TGT_OS_TYPE_VXWORKS || TGT_OS_TYPE_DARWIN
+        address.sin_len = static_cast<U8>(sizeof(struct sockaddr_in));
+#endif
+
+        // First IP address to socket sin_addr
+        if ((status = IpSocket::addressToIp4(m_hostname, &(address.sin_addr))) != SOCK_SUCCESS) {
+            ::close(socketFd);
+            return status;
+        };
+
+        // Now apply timeouts
+        if ((status = IpSocket::setupTimeouts(socketFd)) != SOCK_SUCCESS) {
+            ::close(socketFd);
```

Comment:
Double Close
socketFd was closed at IpSocket.cpp:71 and is closed again here by close().
Might need to check for the close before attempting to close again.

Reply:
Will remove the line IpSocket.cpp:71 to look like other (library) function calls, which do not close socket for you.